mautic/mautic
# Tokens added to cc bcc fields

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | https://github.com/mautic/mautic-documentation/pull/259
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds option to use contact field tokens in the **Send Email to User** campaign action. The tokens used must be based on a field type of `email`. Any other tokens will be rejected by the validation.

#### To Do
- [x] Validate the values for email address (exists now) OR validate them for a token with a field type of email.
- [x] Must work for TO, CC and BCC fields
- [x] If the field does not exist anymore then act like it's a blank value. We talked about validation on custom field delete action and decided not to implement it.
- [x] Comma separated values must still work for email address list.
- [x] Tokens can be also comma-separated. Even mixed comma-separated tokens and email addresses.

#### Steps to test this PR:

##### Test validation
0. Create a new contact email field with alias `email2`.
1. Create a new campaign
2. Add the "Send email to user" action.
3. Save the campaign so you won't have to configure it every time.
4. Try to save different values to the To, BC, BCC fields like
  - `some@valid.email, {invalid-token}` _should trow a validation error that the token is not valid nor an email address_
  - `some@valid.email, {contactfield=firstname}` _should throw a validation error that the field 'firstname' is not type of email_
  - `some@valid.email, {contactfield=email2|invalid-default-email-address}` _should throw a validation error that the default value of the token is not a valid email address_
5. Valid values are for example:
  - `some@valid.email, {contactfield=email2}`
  - `some@valid.email, {contactfield=email2|default@email.address}`
6. Try to save the campaign with some valid values and run it. Test that the tokens are being replaced with the correct values. If the token is evaluated as empty then it should ignore sending the email. But it should send it to all the valid email addresses or correctly replaced token values still.
7. Test also that the campaign action is working well when you unpublish the field used in the tokens or if you completely remove it.

#### Other areas of Mautic that may be affected by the change:
1. It looks like some of the code is used by the form action "send form results" and "send email to user" point action.

#### List deprecations along with the new alternative:
1. None

#### List backwards compatibility breaks:
1. None

#### New classes coverage
| class file path  | coverage  |
|---|---|
| app/bundles/CoreBundle/Exception/InvalidValueException.php  | 100%  |
| app/bundles/CoreBundle/Exception/RecordException.php  | 100%  |
| app/bundles/CoreBundle/Exception/RecordNotFoundException.php  | 100%  |
| app/bundles/CoreBundle/Exception/RecordNotPublishedException.php  | 100%  |
| app/bundles/CoreBundle/Form/DataTransformer/BarStringTransformer.php  | 100%  |
| app/bundles/EmailBundle/Validator/EmailOrEmailTokenList.php  | 100%  |
| app/bundles/EmailBundle/Validator/EmailOrEmailTokenListValidator.php  | 92%  |
| app/bundles/LeadBundle/Exception/InvalidContactFieldTokenException.php  | 100%  |
| app/bundles/LeadBundle/Validator/CustomFieldValidator.php  | 93%  |

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9525"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
